juju/client/connection.py

```diff
@@ -827,47 +865,30 @@ async def _connect_with_redirect(self, endpoints):
+    # (into the self._specified facades) with the max versions that both the client and the controller
     # can negotiate on
     def _build_facades(self, facades_from_connection):
         self.facades.clear()
         for facade in facades_from_connection:
             name = facade['name']
-            # the following attempts to get the best facade version for the
-            # client. The client knows about the best facade versions it speaks,
-            # so in order to be compatible forwards and backwards we speak a
-            # common facade versions.
-            if (name not in client_facades) and (name not in self.specified_facades):
-                # if a facade is required but the client doesn't know about
-                # it, then log a warning.
+            if name in self._specified_facades:
+                client_versions = self._specified_facades[name]
+            elif name in client_facades:
+                client_versions = client_facades[name]
+            elif name in known_unsupported_facades:
+                continue
+            else:
                 log.warning(f'unexpected facade {name} received from the controller')
                 continue
 
-            try:
-                # allow the ability to specify a set of facade versions, so the
-                # client can define the non-conservative facade client pinning.
-                if name in self.specified_facades:
-                    client_versions = self.specified_facades[name]['versions']
-                elif name in client_facades:
-                    client_versions = client_facades[name]['versions']
-
-                controller_versions = facade['versions']
-                # select the max version that both the client and the controller know
-                version = max(set(client_versions).intersection(set(controller_versions)))
-            except ValueError:
-                # this can occur if client_verisons is [1, 2] and controller_versions is [3, 4]
-                # there is just no way to know how to communicate with the facades we're trying to call.
+            controller_versions = facade['versions']
+            candidates = set(client_versions) & set(controller_versions)
+            if not candidates:
                 log.warning(f'unknown common facade version for {name},\n'
                             f'versions known to client : {client_versions}\n'
                             f'versions known to controller : {controller_versions}')
-            except errors.JujuConnectionError:
-                # If the facade isn't with in the local facades then it's not
-                # possible to reason about what version should be used. In this
-                # case we should log the facade was found, but we couldn't
-                # handle it.
-                log.warning(f'unexpected facade {name} found, unable to determine which version to use')
```

Comment:
Was this always unreachable? How would you get a `JujuConnectionError` from a bunch of dictionary lookups and set intersections and `max`? Removing it makes sense.

Reply:
that seems to have been the case, indeed.